### PR TITLE
feat(foundryup): allow multiple installed versions

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
- install option renamed from `-v, --version` to
```
    -i, --install   Install a specific version from built binaries
```
- new options added
```
    -l, --list      List versions installed from built binaries
    -u, --use       Use a specific installed version from built binaries
```
- install multiple versions (with `foundryup -i`) (last one becomes active)
```bash
$ foundryup -i nightly-f089dff1c6c24d1ddf43c7cbefee46ea0197c88f
$ foundryup -i nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
```
- versions are extracted in `~/.foundry/versions/` dir and can be listed with `foundryup -l`
```bash
$ foundryup -l
foundryup: nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
foundryup: - forge 0.2.0 (e5dbb7a 2024-12-03T00:24:42.577356416Z)
foundryup: - cast 0.2.0 (e5dbb7a 2024-12-03T00:24:42.537035487Z)
foundryup: - anvil 0.2.0 (e5dbb7a 2024-12-03T00:24:42.571016376Z)
foundryup: - chisel 0.2.0 (e5dbb7a 2024-12-03T00:24:42.529565979Z)

foundryup: nightly-f089dff1c6c24d1ddf43c7cbefee46ea0197c88f
foundryup: - forge 0.2.0 (f089dff 2024-10-06T00:22:11.758232611Z)
foundryup: - cast 0.2.0 (f089dff 2024-10-06T00:22:11.750238669Z)
foundryup: - anvil 0.2.0 (f089dff 2024-10-06T00:22:11.752640285Z)
foundryup: - chisel 0.2.0 (f089dff 2024-10-06T00:22:11.701654109Z)
```
- use a different installed version
```bash
$ foundryup -u nightly-f089dff1c6c24d1ddf43c7cbefee46ea0197c88f
foundryup: use - forge 0.2.0 (f089dff 2024-10-06T00:22:11.758232611Z)
foundryup: use - cast 0.2.0 (f089dff 2024-10-06T00:22:11.750238669Z)
foundryup: use - anvil 0.2.0 (f089dff 2024-10-06T00:22:11.752640285Z)
foundryup: use - chisel 0.2.0 (f089dff 2024-10-06T00:22:11.701654109Z)
```

### CODE REVIEW

File: foundryup/foundryup [hunk 1]
```diff
@@ -23,6 +24,8 @@ main() {
       -r|--repo)        shift; FOUNDRYUP_REPO=$1;;
       -b|--branch)      shift; FOUNDRYUP_BRANCH=$1;;
       -v|--version)     shift; FOUNDRYUP_VERSION=$1;;
```
Comment (zerosnacks) [inline foundryup/foundryup:26]: should we rename `--version` to `--install`? considering we are already making a breaking change

File: foundryup/foundryup [hunk 2]
```diff
@@ -23,6 +24,8 @@ main() {
...
+      -l|--list)        shift; list;;
```
Comment (zerosnacks) [inline foundryup/foundryup:27]: If you attempt to install an invalid version, like:

```
foundryup --version 59f354c179f4e7f6d7292acb3d068815c79286d1
```

or 

```
foundryup --version https://github.com/foundry-rs/foundry/archive/refs/tags/nightly-aa69ed1e46dd61fbf9d73399396a4db4dd527431.tar.gz
```

it won't exit cleanly meaning when you run `--list` it returns

```
foundryup: https:
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/https:/forge: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/https:/forge --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/https:/cast: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/https:/cast --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/https:/anvil: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/https:/anvil --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/https:/chisel: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/https:/chisel --version
foundryup: - 
foundryup: ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/forge: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/forge --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/cast: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/cast --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/anvil: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/anvil --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/chisel: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/chisel --version
foundryup: - 
foundryup: nightly
foundryup: - forge 0.2.0 (2f698e4 2024-12-13T06:35:25.027879296Z)
foundryup: - cast 0.2.0 (2f698e4 2024-12-13T06:35:25.025215009Z)
foundryup: - anvil 0.2.0 (2f698e4 2024-12-13T06:35:25.023959451Z)
foundryup: - chisel 0.2.0 (2f698e4 2024-12-13T06:35:25.026724926Z)
foundryup: nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
foundryup: - forge 0.2.0 (e5dbb7a 2024-12-03T00:24:42.577356416Z)
foundryup: - cast 0.2.0 (e5dbb7a 2024-12-03T00:24:42.537035487Z)
foundryup: - anvil 0.2.0 (e5dbb7a 2024-12-03T00:24:42.571016376Z)
foundryup: - chisel 0.2.0 (e5dbb7a 2024-12-03T00:24:42.529565979Z)
foundryup: nightly-f089dff1c6c24d1ddf43c7cbefee46ea0197c88f
foundryup: - forge 0.2.0 (f089dff 2024-10-06T00:22:11.758232611Z)
foundryup: - cast 0.2.0 (f089dff 2024-10-06T00:22:11.750238669Z)
foundryup: - anvil 0.2.0 (f089dff 2024-10-06T00:22:11.752640285Z)
foundryup: - chisel 0.2.0 (f089dff 2024-10-06T00:22:11.701654109Z)
foundryup: v59f354c179f4e7f6d7292acb3d068815c79286d1
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/forge: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/forge --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/cast: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/cast --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/anvil: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/anvil --version
foundryup: - 
./foo.sh: line 325: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/chisel: No such file or directory
foundryup: command failed: /home/zerosnacks/.foundry/versions/v59f354c179f4e7f6d7292acb3d068815c79286d1/chisel --version
foundryup: - 
```

Comment (zerosnacks) [reply foundryup/foundryup:27]:
```
~/.foundry/versions$ ls
https:                                           nightly                                           nightly-f089dff1c6c24d1ddf43c7cbefee46ea0197c88f
ightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b  nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b  v59f354c179f4e7f6d7292acb3d068815c79286d1
```

File: foundryup/foundryup [hunk 3]
```diff
@@ -252,6 +259,39 @@ OPTIONS:
 EOF
 }
 
+list() {
+  if [ -d "$FOUNDRY_VERSIONS_DIR" ]; then
+    for VERSION in $FOUNDRY_VERSIONS_DIR/*; do
+      say "${VERSION##*/}"
+      for bin in "${BINS[@]}"; do
+        bin_path="$VERSION/$bin"
+        say "- $(ensure "$bin_path" --version)"
```
Comment (zerosnacks) [inline foundryup/foundryup:268]: if possible, would be nice for the last item to append `\n` readability as in:

```
foundryup: nightly
foundryup: - forge 0.2.0 (2f698e4 2024-12-13T06:35:25.027879296Z)
foundryup: - cast 0.2.0 (2f698e4 2024-12-13T06:35:25.025215009Z)
foundryup: - anvil 0.2.0 (2f698e4 2024-12-13T06:35:25.023959451Z)
foundryup: - chisel 0.2.0 (2f698e4 2024-12-13T06:35:25.026724926Z)

foundryup: nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
foundryup: - forge 0.2.0 (e5dbb7a 2024-12-03T00:24:42.577356416Z)
foundryup: - cast 0.2.0 (e5dbb7a 2024-12-03T00:24:42.537035487Z)
foundryup: - anvil 0.2.0 (e5dbb7a 2024-12-03T00:24:42.571016376Z)
foundryup: - chisel 0.2.0 (e5dbb7a 2024-12-03T00:24:42.529565979Z
```

File: foundryup/foundryup (hunk 1 above)
Comment (grandizzy) [reply foundryup/foundryup]: good point, done in https://github.com/foundry-rs/foundry/pull/9551/commits/bf24bd9155a42c152d290771f0a0496ada8ee46a

File: foundryup/foundryup (hunk 2 above)
Comment (grandizzy) [reply foundryup/foundryup:27]: fixed in https://github.com/foundry-rs/foundry/pull/9551/commits/bf24bd9155a42c152d290771f0a0496ada8ee46a

File: foundryup/foundryup (hunk 3 above)
Comment (grandizzy) [reply foundryup/foundryup:268]: yep, added in https://github.com/foundry-rs/foundry/pull/9551/commits/bf24bd9155a42c152d290771f0a0496ada8ee46a

File: foundryup/foundryup (hunk 2 above)
Comment (zerosnacks) [reply foundryup/foundryup:27]: Nice, confirming this has been fixed

**Review by zerosnacks — DISMISSED** [review]
Tested, works well

File: foundryup/foundryup [hunk 4]
```diff
@@ -137,15 +140,16 @@ main() {
...
 
+    ensure mkdir -p $FOUNDRY_VERSIONS_DIR
     # Download and extract the binaries archive
-    say "downloading latest forge, cast, anvil, and chisel"
+    say "downloading forge, cast, anvil, and chisel for $FOUNDRYUP_TAG version"
     if [ "$PLATFORM" = "win32" ]; then
       tmp="$(mktemp -d 2>/dev/null || echo ".")/foundry.zip"
       ensure download "$BIN_ARCHIVE_URL" "$tmp"
-      ensure unzip "$tmp" -d "$FOUNDRY_BIN_DIR"
+      ensure unzip "$tmp" -d "$FOUNDRY_VERSIONS_DIR/$FOUNDRYUP_TAG"
```
Comment (grandizzy) [inline foundryup/foundryup:149]: have to test this works OK on Win

Comment (grandizzy) [reply foundryup/foundryup:149]: tested and files properly inflated / unzipped, plus also asking  to confirm if you want to replace / skip file for already installed versions

File: foundryup/foundryup [hunk 5]
```diff
@@ -240,7 +251,9 @@ USAGE:
 
 OPTIONS:
     -h, --help      Print help information
-    -v, --version   Install a specific version from built binaries
+    -i, --install   Install a specific version from built binaries
```
Comment (zerosnacks) [inline foundryup/foundryup:254]: nits:

not a regression but running `--install` without comments silently fails (this is expected but ideally should raise an error):

```
foundryup --install
```

ideally `--use` without an argument should display `foundryup: version not installed`, right now it panics with `cp: cannot stat '/home/zerosnacks/.foundry/versions//forge': No such file or directory`

./foundryup --use

**Review by zerosnacks — APPROVED** [review]
Tested on MacOS + Linux, works well

Tiny find: `--install` accepts `0.2.0` as a tag (will turn into `v0.2.0` whereas `--use` does not accept `0.2.0`, only `v0.2.0`. Definitely not a blocker as `v0.2.0` is correct.

**Pull request merged**